README.md
# lens-errors

[HACKAGE](http://hackage.haskell.org/package/lens-errors)

Handling failure deep inside a lens chain intelligently is hard;
You tried to crawl deep into a tree, but it failed and you're not sure why!
This helps with that.

This library doesn't aim to serve every single use case; or provide a TON of 
error combinators, it mostly gives you the tools to write your own.

# Example
Here's what a usage could look like;

Let's say we have a nice tree:

```haskelll
let tree = Node 0 [Node 1 [Node 2 []]]
```

But we want to get the node at the bottom! No problem, we can write a traversal
for that!

```haskell
λ> tree ^.. branches . ix 0 . branches . ix 10 . root
[]
```

Oh; we failed! We can probably guess in this example why the failure occured,
but in larger more complex traversals it can be pretty tough to find sometimes;
and since traversals branch out we might even fail in more than one place at a time!

`lens-errors` gives us
a whole shwack of `fizzle` combinators which will allow us to effectively
'kill' a branch of the traversal with an error message while allowing other
branches to continue to limp along and see if they can succeed.

Why `fizzle` you ask? Words like `fail` or `throw` seem to strong, in this
case only certain branches of the traversal might fizzle out while allowing others to continue.

Let's build a new traversal which reports a nice error if we fail to index
into the child branch we want:

```haskell
let tryIx n = ix n `orFizzleWith` (\xs -> [show n <> " was out of bounds in list: " <> show xs])
```

Here we use `orFizzleWith` which will try to run the given traversal; but
will build and throw a nice error if the traversal returns no elements.

Now that we've added errors to our lens chain the standard lens-running combinators
like `^.` and `%~` won't work any more, but we've got new ones. Let's try `^&..`
which runs a traversal and collects the elements into a list, with nice errors along the way!

```haskell
>>> tree ^&.. branches . tryIx 0 . branches . tryIx 10 . root
(["10 was out of bounds in list: [Node {rootLabel = 2, subForest = []}]"],[])
```

What a great error message! Now we know exactly why we didn't manage to find any
elements there.

You'll notice we actually returned a tuple of type `([String], [Int])`; We return
the errors in the left half and values in the right. Because of the branching factor
of traversals, it's entirely possible for one branch to fail and another to succeed!

Let's see an example of that.

```haskell
>>> [1, 2, 3, 4] ^&.. traversed . fizzleWithUnless (\n -> show n <> " wasn't even!") even
("1 wasn't even!3 wasn't even!", [2, 4])
```

Here we see two interesting things! One is that it returned some elements successfully!
But also, it returned a funny looking error message; looks like our messages got smushed together!

`lens-error` lets you use any error type you want given you follow two conditions:

1. All lenses in the chain must agree on the same error type
2. The error type must be a Monoid.

The error combinators will end up mashing all your errors together, so usually
it's nice to just wrap them all in a list; up to you though!

There are a few more `fizzle` combinators available; hopefully you can build what you need with them :)

For instance we can include failure messages on prisms.

If we use `^&?`; We'll find the first successfull result; otherwise return all
the errors we encountered.

```haskell
>>> let prismError p name = p `orFizzleWith` (\v -> ["Value " <> show v <> " didn't match: " <> name])
>>> let _R = prismError _Right "_Right"

>>> [Left (1 :: Int), Left 2, Right (3 :: Int)] ^&? traversed . _R
Success 3
>>> [Left (1 :: Int), Left 2, Left 3] ^&? traversed . _R
Failure [ "Value Left 1 didn't match: _Right"
        , "Value Left 2 didn't match: _Right"
        , "Value Left 3 didn't match: _Right"]
```

# More examples

For more examples you can reference the [test suite](./test/Spec.hs).


ChangeLog.md
# Changelog for lens-errors

## 0.2.2.0
- Adds new `Fizzler` and `Fizzler'` type aliases for working with fizzlers
- Adds new `fizzler` helper for building fizzlers.



test/Spec.hs
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE RankNTypes #-}
import Control.Lens
import Control.Lens.Error
import Test.Hspec
import Data.Tree
import Data.Tree.Lens

numbers :: (String, [ Int ])
numbers = ("hi", [1, 2, 3, 4])

main :: IO ()
main = hspec $ do
    describe "fizzler" $ do
        let fizzHead :: Fizzler' [String] [Int] Int
            fizzHead = fizzler getter setter
            getter :: [Int] -> Either [String] Int
            getter [] = Left ["empty list!"]
            getter (x:_) = Right x
            setter :: [Int] -> Int -> Either [String] [Int]
            setter (x:_) _ | x < 0 = Left ["refusing to set over negative head"]
            setter (_:xs) x = Right (x : xs)
            setter [] _ = Right []
        it "should get when succesful" $ do
            ("hi", [1, 2, 3, 4]) ^&.. _2 . fizzHead
              `shouldBe` (([], [1]))
        it "should return errors from getter" $ do
            ("hi", []) ^&.. _2 . fizzHead
              `shouldBe` ((["empty list!"], []))
        it "should set when succesful" $ do
            (("hi", [1, 2, 3, 4]) & _2 . fizzHead %&~ (*10))
              `shouldBe` (Success ("hi", [10, 2, 3, 4]))
        it "should return errors from getter when setting iff that fails first" $ do
            (("hi", []) & _2 . fizzHead %&~ (*10))
              `shouldBe` Failure ["empty list!"]
        it "should return errors from setter iff getter passes but setter fails" $ do
            (("hi", [-10, 2, 3, 4]) & _2 . fizzHead %&~ (*10))
              `shouldBe` Failure ["refusing to set over negative head"]

    describe "examine (^&.)" $ do
        it "should view properly through traversals/folds" $ do
            numbers ^&. _2 . traversed . to show
              `shouldBe` ((), "1234")

        it "should view properly through successful assertions" $ do
            numbers ^&. _2 . traversed . fizzleWhen ["shouldn't fail"] (const False) . to show
              `shouldBe` ([], "1234")

        it "should collect failures when they occur" $ do
            numbers ^&. _2 . traversed . fizzleWithWhen (\n -> [show n]) (const True) . to show
              `shouldBe` (["1", "2", "3", "4"], "")

        it "should collect failures AND successes when they occur" $ do
            numbers ^&. _2 . traversed . fizzleWithWhen (\n -> [show n]) even . to (:[])
              `shouldBe` (["2", "4"], [1, 3])

    describe "examineList (^&..)" $ do
        it "should view properly through traversals/folds" $ do
            numbers ^&.. _2 . traversed
              `shouldBe` ((), [1, 2, 3, 4])

        it "should view properly through successful assertions" $ do
            numbers ^&..  (_2 . traversed . fizzleWhen ["shouldn't fail"] (const False))
              `shouldBe` ([], [1, 2, 3, 4])

        it "should collect failures when they occur" $ do
            numbers ^&..  (_2 . traversed . fizzleWithWhen (\n -> [show n]) (const True))
              `shouldBe` (["1", "2", "3", "4"], [])

        it "should collect failures AND successes when they occur" $ do
            numbers ^&..  (_2 . traversed . fizzleWithWhen (\n -> [show n]) even)
              `shouldBe` (["2", "4"], [1, 3])

    describe "preexamine ^&?" $ do
        it "Should find first success or return all errors" $ do
            let prismError p name = p `orFizzleWith` (\v -> ["Value " <> show v <> " didn't match: " <> name])
            let _R = prismError _Right "_Right"
            ([Left (1 :: Int), Left 2, Right (3 :: Int)] ^&? traversed . _R)
              `shouldBe` (Success 3)
            ([Left (1 :: Int), Left 2, Left 3] ^&? traversed . _R)
              `shouldBe` (Failure [ "Value Left 1 didn't match: _Right"
                                  , "Value Left 2 didn't match: _Right"
                                  , "Value Left 3 didn't match: _Right"])
    describe "trySet .&~" $ do
        it "should set successfully" $ do
            (numbers & _2 . ix 1 . fizzleWhen ["shouldn't fail"] (const False) .&~ 42)
              `shouldBe` Success ("hi",[1,42,3,4])
        it "should return failures" $ do
            (numbers & _2 . ix 1 . fizzleWithWhen (\n -> [n]) even .&~ 42)
              `shouldBe` Failure [2]

    describe "tryModify %&~" $ do
        it "should edit successfully with no assertions" $ do
            (numbers & _2 . traversed %&~ (*100))
              `shouldBe` (Success ("hi", [100, 200, 300, 400]) :: Validation () (String, [Int]))
        it "should edit successfully through valid assertions" $ do
            (numbers & _2 . traversed . fizzleWhen ["shouldn't fail"] (const False) %&~ (*100))
              `shouldBe` (Success ("hi", [100, 200, 300, 400]))
        it "should return failures" $ do
            (numbers & _2 . traversed . fizzleWithWhen (\n -> [n]) (const True) %&~ (*100))
              `shouldBe` Failure [1, 2, 3, 4]
        it "should collect all failures if anything fails" $ do
            (numbers & _2 . traversed . fizzleWithWhen (\n -> [n]) even %&~ (*100))
              `shouldBe` Failure [2, 4]

    describe "tryModify' %%&~" $ do
        it "should edit successfully with no assertions" $ do
            (numbers & _2 . traversed %%&~ Success . (*100))
              `shouldBe` (Success ("hi", [100, 200, 300, 400]) :: Validation () (String, [Int]))
        it "should edit successfully through valid assertions" $ do
            (numbers & _2 . traversed . fizzleWhen ["shouldn't fail"] (const False) %%&~ Success . (*100))
              `shouldBe` (Success ("hi", [100, 200, 300, 400]))
        it "should return failures" $ do
            (numbers & _2 . traversed . fizzleWithWhen (\n -> [n]) (const True) %%&~ Success . (*100))
              `shouldBe` Failure [1, 2, 3, 4]
        it "should collect all failures if anything fails" $ do
            (numbers & _2 . traversed . fizzleWithWhen (\n -> [n]) even %%&~ Success . (*100))
              `shouldBe` Failure [2, 4]
        it "should fail if the function fails" $ do
            (numbers & _2 . traversed %%&~ (\n -> Failure [show n <> " failed"]))
              `shouldBe` (Failure ["1 failed","2 failed","3 failed","4 failed"] :: Validation [String] (String, [Int]))

    describe "fizzleWhen" $ do
        it "should fizzle when predicate is true" $ do
            numbers ^&.. _2 . traversed . fizzleWhen ["failure"] even
              `shouldBe` (["failure", "failure"], [1, 3])
    describe "fizzleUnless" $ do
        it "should fizzle when predicate is false" $ do
            numbers ^&.. _2 . traversed . fizzleUnless ["failure"] even
              `shouldBe` (["failure", "failure"], [2, 4])
    describe "maybeFizzleWith" $ do
        it "should fizzle when returning Just" $ do
            let p x
                 | even x    = Just [show x <> " was even"]
                 | otherwise = Nothing
            numbers ^&.. _2 . traversed . maybeFizzleWith p
              `shouldBe` (["2 was even", "4 was even"], [1, 3])
    describe "fizzleWithWhen" $ do
        it "should fizzle using the error builder when predicate is true" $ do
            let p x = [show x <> " was even"]
            numbers ^&.. _2 . traversed . fizzleWithWhen p even
              `shouldBe` (["2 was even", "4 was even"], [1, 3])
    describe "fizzleWithUnless" $ do
        it "should fizzle using the error builder when predicate is false" $ do
            let p x = [show x <> " was even"]
            numbers ^&.. _2 . traversed . fizzleWithUnless p odd
              `shouldBe` (["2 was even", "4 was even"], [1, 3])
    describe "fizzleWith" $ do
        it "should always fizzle using the error builder" $ do
            let p x = [show x]
            numbers ^&.. _2 . traversed . fizzleWith p
              `shouldBe` (["1", "2", "3", "4"], [] :: [Int])
    describe "orFizzle" $ do
        it "should always fizzle using the error builder" $ do
            numbers ^&.. (_2 . traversed . filtered (> 10)) `orFizzle` ["nothing over 10"]
              `shouldBe` (["nothing over 10"], [])
    describe "orFizzleWith" $ do
        it "should always fizzle using the error builder" $ do
            numbers ^&.. (_2 . traversed . filtered (> 10)) `orFizzleWith` (\(_, xs) -> ["searched " <> show (length xs) <> " elements, no luck"])
              `shouldBe` (["searched 4 elements, no luck"], [])
    describe "adjustingErrors" $ do
        it "should alter errors from its sub-branch, but not outside of it" $ do
            [1, 2, 3, 4 :: Int] ^&.. traversed . fizzleWhen ["got 4"] (== 4) . adjustingErrors (fmap (<> "!")) . fizzleWhen ["got 3"] (== 3)
              `shouldBe` (["got 3!", "got 4"], [1, 2])
    describe "adjustingErrorsWith" $ do
        it "should alter errors from its sub-branch, but not outside of it, using the value to construct the error" $ do
            [1, 2, 3, 4 :: Int] ^&.. traversed . fizzleWhen ["got 4"] (== 4) . adjustingErrorsWith (\n -> fmap (\e -> show n <> ": " <> e)) . fizzleWhen ["fail"] (== 3)
              `shouldBe` (["3: fail","got 4"], [1, 2])

    describe "real examples" $ do
        it "tree get success" $ do
            let tree = Node "top" [Node "mid" [Node "bottom" []]]
            let tryIx n = ix n `orFizzleWith` (\xs -> [show n <> " was out of bounds in list: " <> show xs])
            tree ^&.. branches . tryIx 0 . branches . tryIx 0 . root
                `shouldBe` ([],["bottom"])
        it "tree get failure" $ do
            let tree = Node "top" [Node "mid" [Node "bottom" []]]
            let tryIx n = ix n `orFizzleWith` (\xs -> [show n <> " was out of bounds in list: " <> show xs])
            tree ^&.. branches . tryIx 0 . branches . tryIx 10 . root
                `shouldBe` (["10 was out of bounds in list: [Node {rootLabel = \"bottom\", subForest = []}]"],[])
        it "tree set" $ do
            let tree = Node "top" [Node "mid" [Node "bottom" []]] :: Tree String
            let tryIx :: (Applicative f, LensFail [String] f, Show a) => Int -> LensLike' f [a] a
                tryIx n = ix n `orFizzleWith` (\xs -> [show n <> " was out of bounds in list: " <> show xs])
            (tree & branches . tryIx 0 . branches . tryIx 10 . root %&~ (<> "!!"))
                `shouldBe` (Failure ["10 was out of bounds in list: [Node {rootLabel = \"bottom\", subForest = []}]"])




src/Lib.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE DeriveFunctor #-}
module Lib where

import Control.Lens
import Data.Monoid
-- import Data.Either.Validation

class CanFail e f where
  throw :: e -> f a

instance CanFail e (MyValidator e) where
  throw e = Failure e

instance CanFail e (Const (MyValidator e a)) where
  throw e = Const (throw e)

instance Monoid m => CanFail e (Const m) where
  throw _ = Const mempty

data MyValidator e a = Failure e | Success a
  deriving (Show, Eq, Functor)

instance Semigroup e => Applicative (MyValidator e) where
  pure a = Success a
  Failure e <*> Failure e' = Failure (e <> e')
  Failure e <*> Success _ = Failure e
  Success _ <*> Failure e = Failure e
  Success f <*> Success a = Success (f a)

instance (Semigroup e, Semigroup a) => Semigroup (MyValidator e a) where
  Failure e <> Failure e' = Failure (e <> e')
  Success _ <> Failure e = Failure e
  Failure e <> Success _ = Failure e
  Success a <> Success a' = Success (a <> a')

instance (Semigroup e, Monoid a) => Monoid (MyValidator e a) where
  mempty = Success mempty

verify :: CanFail e f => (s -> Maybe e) -> LensLike' f s s
verify check f s =
    case check s of
        Nothing -> f s
        Just e -> throw e

viewSafe :: LensLike (Const (MyValidator e [a])) s t a b -> s -> MyValidator e [a]
viewSafe l s = getConst $ (l %%~ Const . Success . pure) s

overSafe :: (a -> MyValidator e b) -> LensLike (MyValidator e) s t a b -> s -> MyValidator e t
overSafe f l s = (l %%~ f) s

badData :: (String, [Int])
badData = ("hi", [1, 2, 3, 4])

goodData :: (String, [Int])
goodData = ("hi", [2, 4])


traverseEvenNums :: (Applicative f, CanFail [String] f) => LensLike' f (String, [Int]) Int
traverseEvenNums =  _2 . traversed . verify checkEven
  where
    checkEven x | even x = Nothing
                | otherwise = Just [show x <> " is not an even number"]



-- λ> viewSafe traverseEvenNums badData
-- Failure ["1 is not an even number","3 is not an even number"]
-- λ> viewSafe traverseEvenNums goodData
-- Success [2,4]
-- λ> overSafe (Success . (*10)) traverseEvenNums goodData
-- Success ("hi",[20,40])
-- λ> overSafe (Success . (*10)) traverseEvenNums badData
-- Failure ["1 is not an even number","3 is not an even number"]



traverseEvenNums' :: LensLike' (Either String) [Int] Int
traverseEvenNums' f xs = traverse go xs
    where
      go x | even x = f x
           | otherwise = Left (show x <> " is not an even number")

-- λ> ("hi", [1, 2, 3]) & _2 . traverseEvenNums' %%~ pure . (*100)
-- Left "1 is not an even number"
-- λ> ("hi", [2, 4]) & _2 . traverseEvenNums' %%~ pure . (*100)
-- Right ("hi",[200,400])


src/Control/Lens/Error/Internal/LensFail.hs
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleInstances #-}
module Control.Lens.Error.Internal.LensFail where

import Data.Functor.Const
import Data.Either.Validation
import Control.Monad
import Data.Functor.Contravariant

class LensFail e f | f -> e where
  fizzle :: e -> f a
  alterErrors :: (e -> e) -> f a -> f a
  joinErrors :: f (f a) -> f a

instance Monoid a => LensFail e (Const (e, a)) where
  fizzle e = Const (e, mempty)
  alterErrors f (Const (e, a)) = Const (f e, a)
  joinErrors = phantom

instance LensFail e (Const (Either e a)) where
  fizzle e = Const (fizzle e)
  alterErrors f (Const (Left e)) = Const (Left (f e))
  alterErrors _ fa = fa
  joinErrors = phantom

instance LensFail e (Const (Validation e a)) where
  fizzle e = Const (fizzle e)
  alterErrors f (Const (Failure e)) = Const (Failure (f e))
  alterErrors _ fa = fa
  joinErrors = phantom

instance LensFail e (Either e) where
  fizzle e = Left e
  alterErrors f (Left e) = Left (f e)
  alterErrors _ fa = fa
  joinErrors = join

instance LensFail e (Validation e) where
  fizzle e = Failure e
  alterErrors f (Failure e) = Failure (f e)
  alterErrors _ fa = fa
  joinErrors (Failure e) = Failure e
  joinErrors (Success a) = a


src/Control/Lens/Error.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Control.Lens.Error
    (
    -- * Actions
      examine
    , examineList
    , preexamine
    , trySet
    , tryModify
    , tryModify'

    -- * Operators
    , (^&.)
    , (^&..)
    , (^&?)
    , (.&~)
    , (%&~)
    , (%%&~)

    -- * Failing
    , fizzler
    , fizzleWhen
    , fizzleUnless
    , fizzleWith
    , fizzleWithWhen
    , fizzleWithUnless
    , maybeFizzleWith
    , orFizzle
    , orFizzleWith

    -- * Adjusting Errors
    , adjustingErrors
    , adjustingErrorsWith

    -- * Types
    , Fizzler
    , Fizzler'

    -- * Classes
    , LensFail(..)

    -- * Re-exports
    , module Data.Either.Validation
    ) where

import Control.Lens.Error.Internal.LensFail
import Control.Lens
import Data.Either.Validation
import Data.Monoid

-- | Represents a lens-like which may fail with an error of type @e@
type Fizzler e s t a b = forall f. (LensFail e f, Applicative f) => LensLike f s t a b

-- | Represents a simple 'Fizzler'
type Fizzler' e s a = Fizzler e s s a a

-- | Construct a fizzler allowing failure both in the getter and setter.
fizzler :: (s -> Either e a) -> (s -> b -> Either e t) -> Fizzler e s t a b
fizzler viewFizzler setFizzle f s =
    case viewFizzler s of
        Left e -> fizzle e
        Right a -> joinErrors (go <$> f a)
  where
    go b = case setFizzle s b of
             Left e -> fizzle e
             Right t -> pure t

-- | Cause the current traversal to fizzle with a failure when the focus matches a predicate
--
-- >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . fizzleWhen ["failure"] even :: ([String], [Int])
-- (["failure","failure"],[1,3])
fizzleWhen :: e -> (s -> Bool) -> Fizzler' e s s
fizzleWhen e check f s | check s = fizzle e
                       | otherwise = f s

-- | Cause the current traversal to fizzle with a failure when the focus fails a predicate
--
-- >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . fizzleUnless ["failure"] even
-- (["failure","failure"],[2,4])
fizzleUnless :: e -> (s -> Bool) -> Fizzler' e s s
fizzleUnless e check = fizzleWhen e (not . check)

-- | Given a function which might produce an error, fizzle  on 'Just', pass through on 'Nothing'
--
-- > >>> let p x
-- > >>> | even x    = Just [show x <> " was even"]
-- > >>> | otherwise = Nothing
-- > >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . maybeFizzleWith p
-- > (["2 was even","4 was even"],[1,3])
maybeFizzleWith :: (s -> Maybe e) -> Fizzler' e s s
maybeFizzleWith check f s =
    case check s of
        Nothing -> f s
        Just e -> fizzle e

-- | Fizzle using the given error builder when the focus matches a predicate
--
-- >>> let p x = [show x <> " was even"]
-- >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . fizzleWithWhen p even
-- (["2 was even","4 was even"],[1,3])
fizzleWithWhen :: (s -> e) -> (s -> Bool) -> Fizzler' e s s
fizzleWithWhen mkErr check f s
  | check s = fizzle $ mkErr s
  | otherwise = f s

-- | Fizzle using the given error builder when the focus fails a predicate
--
-- >>> let p x = [show x <> " was even"]
-- >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . fizzleWithUnless p odd
-- (["2 was even","4 was even"],[1,3])
fizzleWithUnless :: (s -> e) -> (s -> Bool) -> Fizzler' e s s
fizzleWithUnless mkErr check = fizzleWithWhen mkErr (not . check)

-- | Always fizzle with the given error builder
-- >>> let p x = [show x]
-- >>> ("hi", [1, 2, 3, 4]) ^&.. _2 . traversed . fizzleWith p
-- (["1","2","3","4"],[])
fizzleWith :: (s -> e) -> Fizzler e s t a b
fizzleWith mkErr _ s = fizzle (mkErr s)

-- | Fail with the given error when the provided traversal produces no elements.
--
-- >>> ("hi", [1, 2, 3, 4]) ^&.. (_2 . traversed . filtered (> 10)) `orFizzle` ["nothing over 10"]
-- (["nothing over 10"],[])
orFizzle ::
  (LensFail e f, Applicative f) =>
  Traversing (->) f s t a b -> e -> LensLike f s t a b
orFizzle l e = orFizzleWith l (const e)

-- | Fail using the given error builder when the provided traversal produces no elements.
--
-- >>> ("hi", [1, 2, 3, 4]) ^&.. (_2 . traversed . filtered (> 10)) `orFizzleWith` (\(_, xs) -> ["searched " <> show (length xs) <> " elements, no luck"])
-- (["searched 4 elements, no luck"],[])
orFizzleWith ::
  (LensFail e f, Applicative f) =>
  Traversing (->) f s t a b -> (s -> e) -> LensLike f s t a b
orFizzleWith l mkErr = l `failing` fizzleWith mkErr

infixl 8 ^&.
-- | Operator alias of 'examine'
--
-- View the focus of a lens or traversal over a monoid. Returns the element and the monoidal
-- sum of any errors encountered. Analogous to '^.' with error collection.
--
-- >>> ("hi", [1, 2, 3, 4]) ^&. _2 . traversed . fizzleWithWhen (\n -> [show n]) even . to (:[])
-- (["2","4"],[1,3])
(^&.) :: Monoid e => s -> Getting (e, a) s a -> (e, a)
(^&.) s l = examine l s

infixl 8 ^&..
-- | Operator alias of 'examineList'
--
-- View the focuses of a traversal as a list.
-- Returns the elements and the monoidal sum of any errors encountered. Analogous to '^..'
-- with error collection.
--
-- >>> ("hi", [1, 2, 3, 4]) ^&..  (_2 . traversed . fizzleWithWhen (\n -> [show n]) even)
-- (["2","4"],[1,3])
(^&..) :: Monoid e => s -> Getting (e, [a]) s a -> (e, [a])
(^&..) s l = examineList l s

-- | See also '^&.'
--
-- View the focus of a lens or traversal over a monoid. Returns the element and the monoidal
-- sum of any errors encountered. Analogous to '^.' with error collection.
--
-- >>> examine (_2 . traversed . fizzleWithWhen (\n -> [show n]) even . to (:[])) ("hi", [1, 2, 3, 4])
-- (["2","4"],[1,3])
examine :: Monoid e => Getting (e, a) s a -> s -> (e, a)
examine l = getConst . l (Const . (mempty,))

-- | See also '^&..'
--
-- View the focuses of a traversal as a list.
-- Returns the elements and the monoidal sum of any errors encountered. Analogous to '^..'
-- with error collection.
--
-- >>> examineList ((_2 . traversed . fizzleWithWhen (\n -> [show n]) even)) ("hi", [1, 2, 3, 4])
-- (["2","4"],[1,3])
examineList :: Monoid e => Getting (e, [a]) s a -> s -> (e, [a])
examineList l = getConst . l (Const . (mempty,) . (:[]))

infixl 8 ^&?
-- | Operator alias of 'preexamine'
--
-- Find the first element of a traversal; or return all errors found along the way.
--
-- >>> [1, 2, 3, 4] ^&? traversed . fizzleWhen ["odd"] odd
-- Success 2
--
-- >>> [1, 2, 3, 4] ^&? traversed . fizzleWithWhen (\s -> [show s <> " is too small"]) (<10)
-- Failure ["1 is too small","2 is too small","3 is too small","4 is too small"]
(^&?) :: Monoid e => s -> Getting (e, First a) s a -> Validation e a
(^&?) s l = unpack . getConst .  l (Const . (mempty,) . First . Just) $ s
  where
    unpack (_, First (Just a)) = Success a
    unpack (e, First (Nothing)) = Failure e

-- | See also '^&?'
--
-- Find the first element of a traversal; or return all errors found along the way.
--
-- >>> preexamine (traversed . fizzleWhen ["odd"] odd) [1, 2, 3, 4]
-- Success 2
--
-- >>> preexamine (traversed . fizzleWithWhen (\s -> [show s <> " is too small"]) (<10)) [1, 2, 3, 4]
-- Failure ["1 is too small","2 is too small","3 is too small","4 is too small"]
preexamine :: Monoid e => Getting (e, First a) s a -> s -> Validation e a
preexamine l s = s ^&? l

infixl 8 .&~
-- | Operator alias of 'trySet
--
-- Set the focus of a lens/traversal. Returns a monoidal summary of failures or the altered
-- structure.
--
-- >>> ("hi", [1, 2, 3, 4]) & _2 . ix 1 . fizzleWhen ["shouldn't fail"] (const False) .&~ 42
-- Success ("hi",[1,42,3,4])
--
-- >>> ("hi", [1, 2, 3, 4]) & _2 . ix 1 . fizzleWithWhen (\n -> [n]) even .&~ 42
-- Failure [2]
(.&~) :: LensLike (Validation e) s t a b -> b -> s -> Validation e t
(.&~) l b s = s & l %%~ Success . const b

-- | See also '.&~'
--
-- Set the focus of a lens/traversal. Returns a monoidal summary of failures or the altered
-- structure.
--
-- >>> trySet (_2 . ix 1 . fizzleWhen ["shouldn't fail"] (const False)) 42 ("hi", [1, 2, 3, 4])
-- Success ("hi",[1,42,3,4])
--
-- >>> trySet  (_2 . ix 1 . fizzleWithWhen (\n -> [n]) even) 42 ("hi", [1, 2, 3, 4])
-- Failure [2]
trySet :: LensLike (Validation e) s t a b -> b -> s -> Validation e t
trySet = (.&~)

infixl 8 %&~
-- | Operator alias of 'tryModify'
--
-- Modify the focus of a lens/traversal. Returns a monoidal summary of failures or the altered
-- structure.
--
-- >>> ("hi", [1, 2, 3, 4]) & _2 . traversed . fizzleWhen ["shouldn't fail"] (const False) %&~ (*100)
-- Success ("hi",[100,200,300,400])
--
-- >>> ("hi", [1, 2, 3, 4]) & _2 . traversed . fizzleWithWhen (\n -> [n]) even %&~ (*100)
-- Failure [2,4]
(%&~) :: LensLike (Validation e) s t a b -> (a -> b) -> s -> Validation e t
(%&~) l f s = s & l %%~ Success . f

-- | See also '%&~'
--
-- Modify the focus of a lens/traversal. Returns a monoidal summary of failures or the altered
-- structure.
--
-- >>> tryModify  (_2 . traversed . fizzleWhen ["shouldn't fail"] (const False)) (*100) ("hi", [1, 2, 3, 4])
-- Success ("hi",[100,200,300,400])
--
-- >>> tryModify  (_2 . traversed . fizzleWithWhen (\n -> [n]) even) (*100) ("hi", [1, 2, 3, 4])
-- Failure [2,4]
tryModify :: LensLike (Validation e) s t a b -> (a -> b) -> s -> Validation e t
tryModify l f s = s & l %&~  f

infixl 8 %%&~
-- | Operator alias of 'tryModify''
--
-- Modify the focus of a lens/traversal with a function which may fail.
-- Returns a monoidal summary of failures or the altered structure.
--
-- The definition of this function is actually just:
--
-- > (%%&~) = (%%~)
--
-- But this combinator is provided for discoverability, completeness, and hoogle-ability.
--
-- >>> ("hi", [1, 2, 3, 4]) & _2 . traversed . fizzleWithWhen (\n -> [n]) even %%&~ Success . (*100)
-- Failure [2,4]
-- >>> ("hi", [1, 2, 3, 4]) & _2 . traversed %%&~ (\n -> Failure [show n <> " failed"])
-- Failure ["1 failed","2 failed","3 failed","4 failed"]
(%%&~) :: LensLike (Validation e) s t a b -> (a -> Validation e b) -> s -> Validation e t
(%%&~) = (%%~)

-- | See also '%%&~'
--
-- Modify the focus of a lens/traversal with a function which may fail.
-- Returns a monoidal summary of failures or the altered structure.
--
-- >>> tryModify' (_2 . traversed . fizzleWithWhen (\n -> [n]) even) (Success . (*100)) ("hi", [1, 2, 3, 4])
-- Failure [2,4]
-- >>> tryModify' (_2 . traversed) (\n -> Failure [show n <> " failed"]) ("hi", [1, 2, 3, 4])
-- Failure ["1 failed","2 failed","3 failed","4 failed"]
tryModify' :: LensLike (Validation e) s t a b -> (a -> Validation e b) -> s -> Validation e t
tryModify' l f s = s & l %%~ f


-- | Adjust any errors which occur in the following branch.
-- Note that we can't change the error type, but this can be helpful for adding context
-- to errors if they occur at a position without enough context.
--
-- This is does nothing when no errors occur.
--
-- >>> [1, 2, 3, 4 :: Int] ^&.. traversed . fizzleWhen ["got 4"] (== 4) . adjustingErrors (fmap (<> "!")) . fizzleWhen ["got 3"] (== 3)
-- (["got 3!","got 4"],[1,2])
adjustingErrors :: (e -> e) -> Fizzler' e s s
adjustingErrors addCtx f s = alterErrors addCtx (f s)

-- | Adjust any errors which occur in the following branch, using the value available at
-- the current position to add context..
-- Note that we can't change the error type, but this can be helpful for adding context
-- to errors if they occur at a position without enough context.
--
-- This is does nothing when no errors occur.
--
-- >>> [1, 2, 3, 4 :: Int] ^&.. traversed . fizzleWhen ["got 4"] (== 4) . adjustingErrorsWith (\n -> fmap (\e -> show n <> ": " <> e)) . fizzleWhen ["fail"] (== 3)
-- (["3: fail","got 4"],[1,2])
adjustingErrorsWith :: (s -> e -> e) -> Fizzler' e s s
adjustingErrorsWith addCtx f s = alterErrors (addCtx s) (f s)
